File: leetcode/unique_binary_search_treesII.js
/* Given n, generate all structurally unique BST's (binary search trees) that 
 * store values 1...n.
 * Time Complexity: O(n4^n)
 * Space Complexity: O(n4^n)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var generateTrees = function(n) {
    if (n === 0)
        return [];
    var DP = new Array(n + 2);
    for (var i = 0; i < DP.length; ++i) {
        for (var temp = new Array(n + 2), j = 0; j < n + 2; ++j) 
            temp[j] = [null];
        DP[i] = temp;
    }
    for (var k = 0; k < n; ++k) {
        for (i = 1; i < n - k + 1; ++i) {
            DP[i][i + k] = [];
            for (var root = i; root < i + k + 1; ++root) {
                for (var left of DP[i][root - 1]) {
                    for (var right of DP[root + 1][i + k]) {
                        var node = new TreeNode(root);
                        node.left = left;
                        node.right = right;
                        DP[i][i + k].push(node);
                    }
                }
            }
        }
    }
    return DP[1][n];
};

File: lintcode/unique_binary_search_treesII.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    # @paramn n: An integer
    # @return: A list of root
    def generateTrees(self, n):
        if n == 0: return [None]
        DP = [[[None] for _ in xrange(n + 2) ] for _ in xrange(n + 2)]
        for k in xrange(n):
            for i in xrange(1, n - k + 1):
                DP[i][i + k] = []
                for root in xrange(i, i + k + 1):
                    for left in DP[i][root - 1]:
                        for right in DP[root + 1][i + k]:
                            node = TreeNode(root)
                            node.left, node.right = left, right
                            DP[i][i + k].append(node)
        return DP[1][n]
